<a href="https://colab.research.google.com/github/Idowuilekura/moodle_analysis/blob/master/10academypostgresFINALSUBMISSION.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Moodle Database: Educational Data Log Analysis
Many educational facilities such as colleges, universities, bootcamps rely on scalable and versatile Learning Management Systems. 
The Moodle LMS  is a free and open-source learning management system written in PHP and distributed under the GNU General Public License. It is used for blended learning, distance education, flipped classroom and other e-learning projects in schools, universities, workplaces and other sectors. With customizable management features, it is used to create private websites with online courses for educators and trainers to achieve learning goals. Moodle allows for extending and tailoring learning environments using community-sourced plugins.

Many educational facilities such as colleges, universities, bootcamps rely on scalable and versatile Learning Management Systems. 

In 2019, 10 Academy( an educational company with the goal of rasing 10x datascientist in Africa) used the Moodle LMS to manage about 1000 students in their 6 months data science training. Learners, course instructors, and all admins interacted with the Moodle system for almost all the training activities. All events from these activities are logged in the moodle postgres database.

For this task 10academy hired me to analyze the dataset



The tasks are divided into four sub-tasks
- Moodle database schema understanding
- Data Extraction Transformation and Loading (ETL)
- Dashboard making with Tableau
- Automation and Deployment 



The columns of intrest in this analysis are 
- **mdl_logstore_standard_log
- mdl_context
- mdl_user
- mdl_course
- mdl_modules 
- mdl_course_modules
- mdl_course_modules_completion**
- mdl_grade_items
- **mdl_grade_grades**
- mdl_grade_categories
- mdl_grade_items_history
- mdl_grade_grades_history
- mdl_grade_categories_history
- mdl_forum
- mdl_forum_discussions
- mdl_forum_posts


In [ ]:
# loading relevant modules 
%load_ext sql
from sqlalchemy import create_engine #used to store our queries as pandas dataframe
%sql postgresql://Samsung:ikujesumuyiwa96@localhost/moodle
import matplotlib.pyplot as plt
%matplotlib inline 
import pandas as pd
import datetime as datetime
import numpy as np
import seaborn as sns
sns.set_style('darkgrid')

In [ ]:
engine = create_engine('postgresql://Samsung:*******@localhost/moodle')

In [ ]:
%%sql 
-- testing if the connection is great
SELECT *
FROM mdl_logstore_standard_log 
LIMIT 3

 * postgresql://Samsung:***@localhost/moodle
3 rows affected.


[(1, '\\core\\event\\course_viewed', 'core', 'viewed', 'course', None, None, 'r', 2, 2, 50, 1, 0, 1, None, 0, 'N;', 1548311114, 'web', '105.233.48.139', None),
 (2, '\\core\\event\\user_loggedin', 'core', 'loggedin', 'user', 'user', 2, 'r', 0, 1, 10, 0, 2, 0, None, 0, 'a:1:{s:8:"username";s:5:"admin";}', 1548311148, 'web', '105.233.48.139', None),
 (3, '\\core\\event\\dashboard_viewed', 'core', 'viewed', 'dashboard', None, None, 'r', 0, 5, 30, 2, 2, 0, 2, 0, 'N;', 1548311179, 'web', '105.233.48.139', None)]

In [ ]:
# testing with sqlachemy pandas object
log_df = pd.read_sql("SELECT * FROM mdl_logstore_standard_log LIMIT 3",engine)
log_df

id                     eventname component    action     target  \
0   1     \core\event\course_viewed      core    viewed     course   
1   2     \core\event\user_loggedin      core  loggedin       user   
2   3  \core\event\dashboard_viewed      core    viewed  dashboard   

  objecttable  objectid crud  edulevel  contextid  ...  contextinstanceid  \
0        None       NaN    r         2          2  ...                  1   
1        user       2.0    r         0          1  ...                  0   
2        None       NaN    r         0          5  ...                  2   

   userid  courseid  relateduserid  anonymous  \
0       0         1            NaN          0   
1       2         0            NaN          0   
2       2         0            2.0          0   

                               other timecreated  origin              ip  \
0                                 N;  1548311114     web  105.233.48.139   
1  a:1:{s:8:"username";s:5:"admin";}  1548311148     web  105.233.48.139   
2                                 N;  1548311179     web  105.233.48.139   

  realuserid  
0       None  
1       None  
2       None  

[3 rows x 21 columns]

## Task 1 - Moodle Database schema understanding

The first thing before doing any analysis is to understand the database, this is a fundamental thing prior to doing any analyis, as it makes whoever is doing analysis to know how to tackle the required problem.To understand the database, we will do these 
- Count the number of tables in the database 
- The number of records in each of the tables given in the relevant tables
- Number of quiz submission by hour of day 
- Monthly usage time of learners who have confirmed and are not deleted 
- Count of log events per user for the following verbs: [loggedin,viewed,started,submitted,uploaded,updated,searched,resumed,answered,attempted,abandoned]

## Task 1(i) Count of number of tables 

In [ ]:
%%sql 
-- counting the number of tables in the database
SELECT COUNT(*) AS NUMBER_OF_TABLES
FROM INFORMATION_SCHEMA.TABLES;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(636,)]

From here we could see that the number of tables in the database is 636( quite much)

## Task(ii) Number of records in each of the tables in the Most important tables 

In [ ]:
%%sql 
-- checking the number of records in mdl_logstore_standard_log
SELECT COUNT(*) as Number_of_records
FROM mdl_logstore_standard_log;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(417554,)]

In [ ]:
%%sql 
-- checking the number of records in mdl_context
SELECT COUNT(*) AS Number_of_records
FROM mdl_user;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(1052,)]

In [ ]:
%%sql 
-- checking the number of records in mdl_course 
SELECT COUNT(*) AS Number_of_records
FROM mdl_course;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(15,)]

In [ ]:
%%sql 
-- checking the number of records in mdl_modules
SELECT COUNT(*) AS Number_of_records
FROM mdl_modules;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(26,)]

In [ ]:
%%sql 
--checking the number of records in mdl_course_modules
SELECT COUNT(*) AS Number_of_records
FROM mdl_course_modules;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(290,)]

In [ ]:
%%sql 
--checking the number of records in course_modules_completion
SELECT COUNT(*) AS Number_of_records
FROM mdl_course_modules_completion;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(4483,)]

In [ ]:
%%sql 
-- checking the number of records in mdl_grade_grade
SELECT COUNT(*) AS Number_of_records
FROM mdl_grade_grades;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(3643,)]

From here we can see that the mdl_logstore_standard_log has the highest number of records with about 417554, well it is not suprising since that logs all the activites done on moodle 

The codes below are basically to check the contents and the colum names in each respective columns

In [ ]:
%%sql 
SELECT * 
FROM mdl_logstore_standard_log
limit 1;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(1, '\\core\\event\\course_viewed', 'core', 'viewed', 'course', None, None, 'r', 2, 2, 50, 1, 0, 1, None, 0, 'N;', 1548311114, 'web', '105.233.48.139', None)]

In [ ]:
%%sql 
select *
FROM mdl_context
limit 1;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(1, 10, 0, '/1', 1, 0)]

In [ ]:
%%sql 
select *
FROM mdl_user
limit 5;

 * postgresql://Samsung:***@localhost/moodle
5 rows affected.


[(1, 1, 0, 0, 0, '', 'Germany', 'en', 0, 0, 0, 0, '', 'Male', 'DE'),
 (20, 1, 0, 0, 0, '', 'Germany', 'en', 1549745767, 1562935333, 1549745767, 1562935333, '197.215.32.111', 'Male', 'DE'),
 (15, 1, 0, 0, 0, '', 'Germany', 'en', 1549633236, 1563537002, 1563469481, 1563537002, '41.184.239.200', 'Male', 'DE'),
 (36, 1, 0, 0, 0, '', 'Ghana', 'en', 1549641288, 1559227964, 1549651186, 1559227892, '154.160.1.224', 'Male', 'GH'),
 (44, 1, 0, 0, 0, '', 'Ghana', 'en', 1549637001, 1560095090, 1557475553, 1560094908, '129.122.1.14', 'Male', 'GH')]

In [ ]:
%%sql 
SELECT * 
FROM mdl_modules
limit 1;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(2, 'assignment', 60, 0, '', 0)]

In [ ]:
%%sql 
SELECT *
FROM mdl_course
limit 1;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(7, 6, 60002, 'FDP Design -WordPress', 'FDP WordPress', '', '<p><img src="@@PLUGINFILE@@/WordPress.svg.png" alt="WordPress website" role="presentation" class="img-responsive atto_image_button_text-bottom" heigh ... (604 characters truncated) ... e. You are free to use, distribute and modify it, including for commercial purposes, provided you acknowledge the source and share-alike.<br><br></p>', 1, 'topics', 1, 5, 1528741800, 1560277800, 0, 2097152, 0, 0, 1, 1, 0, 0, 0, '', '', '', 1528706177, 1548697122, 0, 1, 0, 1561503107)]

In [ ]:
%%sql 
SELECT *
FROM mdl_course_modules
limit 1;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(1, 1, 9, 1, 2, None, 1548435669, 0, 0, 1, 1, 1, 0, 0, 0, None, 0, 0, 0, None, 0)]

In [ ]:
%%sql 
SELECT * 
FROM mdl_course_modules_completion 
limit 1;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(5801, 224, 56, 1, 0, None, 1561988963)]

In [ ]:
%%sql 
SELECT * 
FROM mdl_grade_grades
limit 5;

 * postgresql://Samsung:***@localhost/moodle
5 rows affected.


[(2, 92, 3, None, Decimal('100.00000'), Decimal('0.00000'), None, None, None, 0, 0, 0, 0, 0, 0, None, 0, None, 0, None, None, 'unknown', None),
 (1, 93, 3, None, Decimal('10.00000'), Decimal('0.00000'), None, 3, None, 0, 0, 0, 0, 0, 0, None, 0, None, 0, None, None, 'novalue', Decimal('0.00000')),
 (51, 126, 50, None, Decimal('12.00000'), Decimal('0.00000'), None, None, Decimal('12.00000'), 0, 0, 0, 0, 0, 0, None, 0, None, 0, None, 1549723271, 'unknown', None),
 (35, 92, 65, None, Decimal('12.00000'), Decimal('0.00000'), None, None, Decimal('0.00000'), 0, 0, 0, 0, 0, 0, None, 0, None, 0, None, 1549642424, 'unknown', None),
 (34, 93, 65, Decimal('0.00000'), Decimal('12.00000'), Decimal('0.00000'), None, 3, Decimal('0.00000'), 0, 0, 0, 0, 0, 0, None, 0, None, 0, 1549639813, 1549642424, 'used', Decimal('1.00000'))]

## 1(iii) Number of quiz submission by hour of day

In [ ]:
"""used pandas read_sql to store the sql query as a dataframe,the code below selected component,action and extracted the hour from
the timecreated if it mathes mod_quiz and it is submitted"""
df = pd.read_sql("SELECT component,action,timecreated,EXTRACT(HOUR FROM to_timestamp(timecreated)) AS HOUR  FROM mdl_logstore_standard_log WHERE component = 'mod_quiz' AND action = 'submitted'",engine)

In [ ]:
""" used pandas group by the count each submission by hour"""
df_2=df.groupby('hour').agg({"action":'count'}).reset_index()
df_2.columns = ['hour','number of submissions']
df_2

hour  number of submissions
0    0.0                     29
1    1.0                      7
2    2.0                     14
3    3.0                     10
4    4.0                     18
5    5.0                     17
6    6.0                     34
7    7.0                     46
8    8.0                     69
9    9.0                     95
10  10.0                     88
11  11.0                    104
12  12.0                     90
13  13.0                     80
14  14.0                     68
15  15.0                     68
16  16.0                     70
17  17.0                     57
18  18.0                     55
19  19.0                     46
20  20.0                     53
21  21.0                     26
22  22.0                     23
23  23.0                     30

In [ ]:
%%sql
SELECT userid,action,COUNT(action) AS number_of_actions
FROM mdl_logstore_standard_log
where userid >=0
GROUP BY userid,action
ORDER BY userid;

 * postgresql://Samsung:***@localhost/moodle
6923 rows affected.


[(0, 'updated', 649),
 (0, 'viewed', 5305),
 (0, 'sent', 846),
 (0, 'failed', 457),
 (1, 'viewed', 2),
 (1, 'loggedin', 1),
 (2, 'assigned', 9493),
 (2, 'completed', 304),
 (2, 'awarded', 1209),
 (2, 'added', 4344),
 (2, 'duplicated', 4),
 (2, 'graded', 43),
 (2, 'sent', 8025),
 (2, 'loggedout', 11),
 (2, 'becameoverdue', 5),
 (2, 'ended', 3),
 (2, 'indexed', 490),
 (2, 'downloaded', 1),
 (2, 'previewed', 5),
 (2, 'loggedin', 169),
 (2, 'shown', 6),
 (2, 'enabled', 11),
 (2, 'started', 14),
 (2, 'restored', 5),
 (2, 'abandoned', 5),
 (2, 'archived', 1),
 (2, 'unassigned', 373),
 (2, 'deleted', 511),
 (2, 'reset', 25),
 (2, 'created', 12175),
 (2, 'uploaded', 21),
 (2, 'reviewed', 10),
 (2, 'submitted', 21),
 (2, 'removed', 1),
 (2, 'failed', 2683),
 (2, 'loggedinas', 6),
 (2, 'disabled', 6),
 (2, 'viewed', 2492),
 (2, 'updated', 2551),
 (3, 'assigned', 2026),
 (3, 'added', 140),
 (3, 'started', 17),
 (3, 'reviewed', 176),
 (3, 'created', 3202),
 (3, 'failed', 1),
 (3, 'reset', 7),
 (3, 'enabled', 8),
 (3, 'viewed', 4041),
 (3, 'saved', 3),
 (3, 'deleted', 59),
 (3, 'text', 10),
 (3, 'updated', 1467),
 (3, 'ended', 3),
 (3, 'moved', 6),
 (3, 'submitted', 2),
 (3, 'sent', 565),
 (3, 'loggedout', 12),
 (3, 'uploaded', 6),
 (3, 'graded', 715),
 (3, 'loggedin', 107),
 (3, 'disabled', 2),
 (3, 'awarded', 316),
 (3, 'exported', 7),
 (3, 'resumed', 4),
 (3, 'assessed', 8),
 (3, 'shown', 7),
 (3, 'previewed', 5),
 (4, 'failed', 3),
 (5, 'graded', 3),
 (5, 'loggedin', 54),
 (5, 'answered', 2),
 (5, 'reviewed', 5),
 (5, 'created', 3),
 (5, 'loggedout', 10),
 (5, 'started', 11),
 (5, 'updated', 19),
 (5, 'uploaded', 1),
 (5, 'viewed', 694),
 (5, 'called', 238),
 (5, 'failed', 4),
 (5, 'resumed', 2),
 (5, 'shown', 7),
 (5, 'ended', 3),
 (5, 'sent', 17),
 (5, 'submitted', 5),
 (5, 'added', 1),
 (7, 'ended', 2),
 (7, 'loggedin', 3),
 (7, 'loggedout', 1),
 (7, 'failed', 2),
 (7, 'viewed', 81),
 (7, 'updated', 1),
 (7, 'started', 2),
 (7, 'shown', 2),
 (8, 'viewed', 44),
 (8, 'started', 1),
 (8, 'shown', 6),
 (8, 'updated', 1),
 (8, 'ended', 1),
 (8, 'loggedin', 2),
 (9, 'restarted', 4),
 (9, 'graded', 4),
 (9, 'shown', 2),
 (9, 'resumed', 1),
 (9, 'awarded', 1),
 (9, 'submitted', 8),
 (9, 'failed', 3),
 (9, 'ended', 7),
 (9, 'reviewed', 2),
 (9, 'started', 11),
 (9, 'viewed', 359),
 (9, 'updated', 16),
 (9, 'answered', 4),
 (9, 'saved', 1),
 (9, 'loggedin', 13),
 (12, 'viewed', 76),
 (13, 'viewed', 60),
 (14, 'viewed', 64),
 (15, 'started', 2),
 (15, 'sent', 43),
 (15, 'loggedin', 5),
 (15, 'viewed', 85),
 (15, 'ended', 2),
 (15, 'failed', 1),
 (15, 'updated', 2),
 (15, 'shown', 7),
 (16, 'reset', 1),
 (16, 'shown', 10),
 (16, 'reviewed', 1),
 (16, 'updated', 3),
 (16, 'ended', 3),
 (16, 'submitted', 1),
 (16, 'loggedin', 5),
 (16, 'viewed', 108),
 (16, 'sent', 295),
 (16, 'started', 6),
 (17, 'viewed', 63),
 (18, 'viewed', 74),
 (19, 'started', 4),
 (19, 'updated', 3),
 (19, 'viewed', 90),
 (19, 'shown', 7),
 (19, 'failed', 1),
 (19, 'called', 26),
 (19, 'ended', 2),
 (19, 'sent', 2),
 (19, 'created', 1),
 (19, 'loggedin', 2),
 (20, 'ended', 1),
 (20, 'started', 2),
 (20, 'loggedin', 2),
 (20, 'shown', 6),
 (20, 'reviewed', 1),
 (20, 'viewed', 78),
 (20, 'graded', 1),
 (20, 'failed', 2),
 (20, 'loggedout', 1),
 (20, 'submitted', 1),
 (21, 'viewed', 70),
 (22, 'viewed', 80),
 (22, 'loggedin', 1),
 (22, 'shown', 1),
 (22, 'ended', 1),
 (22, 'started', 1),
 (23, 'viewed', 67),
 (24, 'loggedin', 3),
 (24, 'updated', 1),
 (24, 'shown', 3),
 (24, 'ended', 2),
 (24, 'viewed', 85),
 (24, 'failed', 4),
 (24, 'loggedout', 1),
 (24, 'started', 3),
 (25, 'submitted', 5),
 (25, 'restarted', 5),
 (25, 'added', 2),
 (25, 'deleted', 1),
 (25, 'sent', 20),
 (25, 'viewed', 711),
 (25, 'shown', 9),
 (25, 'updated', 17),
 (25, 'ended', 7),
 (25, 'answered', 6),
 (25, 'created', 3),
 (25, 'failed', 5),
 (25, 'started', 11),
 (25, 'loggedin', 46),
 (25, 'reviewed', 4),
 (25, 'graded', 1),
 (25, 'saved', 4),
 (25, 'resumed', 10),
 (25, 'called', 1021),
 (25, 'logg

In [ ]:
%%sql 
SELECT * 
FROM mdl_forum_posts
limit 2;

 * postgresql://Samsung:***@localhost/moodle
2 rows affected.


[(1, 1, 0, 2, 1548435938, 1548444475, 1, 'Deployment', '<p>This website is still in testing mode. Public release date: 28 Jan 2019! Stay tuned!</p><p><span class="edited">(Edited by <a href="https://moodle.10academy.org/user/view.php?id=3&amp;course=1">Arun Sharma</a> - original submission Friday, 25 January 2019, 6:05 PM)</span></p>', 1, 0, '', 0, 0, 0),
 (3, 3, 0, 2, 1549622897, 1549622897, 1, '2019 Potential Fellows', '<p>We are happy to announce that this platform is officially kick-started, and the 2017 10 Academy Fellows and the the first batch of the 2019 Potential Fellows are included in the system.&nbsp;<br></p><p>Exciting future ahead!</p>', 1, 0, '', 0, 0, 0)]

In [ ]:
%%sql 
-- Monthly usage of time of learners who have confirmed and are not deleted
SELECT EXTRACT(MONTH FROM to_timestamp(firstaccess)),SUM(EXTRACT(HOUR FROM to_timestamp(lastaccess-firstaccess)))
FROM mdl_user
WHERE confirmed = 1 AND deleted=0
GROUP BY EXTRACT(MONTH FROM to_timestamp(firstaccess));

 * postgresql://Samsung:***@localhost/moodle
7 rows affected.


[(3.0, 513.0),
 (7.0, 171.0),
 (1.0, 27.0),
 (2.0, 2290.0),
 (6.0, 204.0),
 (4.0, 925.0),
 (5.0, 638.0)]

## Task 1(vi) Writing a python class to pull the overall grade of learners and the Number of forum posts and the monthly usage time of learners on the platform

In [ ]:
""" THis class is meant to exract the montly_usage_time,number_of_forum_posts and overall_grade by running the specified queries"""
class grade_forum_post:
    def __init__(self):
        #self.grade = self.overall_grade()
        pass
    def monthly_usage_time(self):
        query = "SELECT EXTRACT(MONTH FROM to_timestamp(firstaccess)) AS HOUR,SUM(EXTRACT(HOUR FROM to_timestamp(lastaccess-firstaccess))) FROM mdl_user WHERE confirmed = 1 AND deleted = 0 GROUP BY EXTRACT(MONTH FROM to_timestamp(firstaccess)) ORDER BY(EXTRACT(MONTH FROM to_timestamp(firstaccess)))"
        self.df_month = pd.read_sql(query,engine)
        return self.df_month
    def number_of_forum_posts(self):
        self.df_forum = pd.read_sql("SELECT COUNT(*) AS Number_of_forum_posts FROM mdl_forum_posts;",engine)
        return self.df_forum
    def overall_grade(self):
        self.df = pd.read_sql("SELECT userid,SUM(finalgrade) AS overall_grade FROM mdl_grade_grades GROUP BY userid ORDER BY userid ",engine)
        return self.df

In [ ]:
number_of_forum_posts = grade_forum_post()
#number_of_forum_posts.number_of_forum_posts()
number_of_forum_posts.monthly_usage_time()

hour     sum
0   1.0    27.0
1   2.0  2290.0
2   3.0   513.0
3   4.0   925.0
4   5.0   638.0
5   6.0   204.0
6   7.0   171.0

In [ ]:
overall_grade_of_users = grade_forum_post()
overall_grade_of_users.overall_grade()

userid  overall_grade
0         2            NaN
1         3      200.00000
2         5      400.66666
3         7      200.00000
4         9      636.00000
..      ...            ...
402     991      200.00000
403    1019      200.00000
404    1025     2700.00000
405    1027     3124.00000
406    1041      204.00000

[407 rows x 2 columns]

In [ ]:
df = pd.read_sql("SELECT * FROM mdl_logstore_standard_log where userid >= 0",engine)

In [ ]:
df2 = pd.read_sql("SELECT id,EXTRACT(HOUR FROM(to_timestamp(lastaccess-firstaccess))) AS HOUR FROM mdl_user",engine)

In [ ]:
np.unique(df['action'])

array(['abandoned', 'accepted', 'added', 'answered', 'archived',
       'assessed', 'assigned', 'awarded', 'becameoverdue', 'called',
       'completed', 'created', 'deleted', 'disabled', 'downloaded',
       'duplicated', 'enabled', 'ended', 'exported', 'failed', 'graded',
       'indexed', 'loggedin', 'loggedinas', 'loggedout', 'moved',
       'previewed', 'printed', 'removed', 'reset', 'restarted',
       'restored', 'resumed', 'reviewed', 'saved', 'searched', 'sent',
       'shown', 'started', 'submitted', 'text', 'unassigned', 'updated',
       'uploaded', 'viewed'], dtype=object)

In [ ]:
df[]

In [ ]:
"""To compute the 1%,5%,10% and 25% of the login and activities of users, i wrote this class below, i will
like to know possible ways to improve this or any correction, especially the rank_percent function though it worked but will
like to know if the approach seems great"""
class login_activity_count:
    def __init__(self,df,df2):
        self.df = df
        self.df2 = df2
    def dedication(self):
        df_2 = self.df2
        df_grouped_2 = df_2.groupby('id').agg({'hour':'sum'}).reset_index()
        #bin_labels = ['25%','10%','5%','1%']
        #df_grouped_2['dedication_percent_group'] = pd.qcut(df_grouped_2['hour'],q=4,labels=bin_labels)
        return df_grouped
    def login_activity_count(self):
        df_1= self.df
        """ wrote a function to check if the row contains action as loggedin, if true assign 1 if false assign 0
        ,then wrote another function to check if action is not equal to loggded,a better way to check for other activities
        """
        df_1['logged_yes'] = df_1['action'].apply(lambda x:1 if (x=='loggedin') else 0)
        df_1['activity_yes'] = df_1['action'].apply(lambda x:1 if x !='loggedin' else 0)
        """ after this the dataframe was grouped by user and summed on both logged_yes and activity_yes
        (which is similar to counting it)"""
        df_grouped = df_1.groupby('userid').agg({"logged_yes":'sum',"activity_yes":"sum"}).reset_index()
        """ next the column was ranked based on the activity_yes and logged_yes number"""
        df_grouped['activity_rank'] = df_grouped['activity_yes'].rank(ascending=False)
        df_grouped['logged_rank'] = df_grouped['logged_yes'].rank(ascending =False)
        length_df = len(df_grouped) #finding the rank of the group
        """ for this calculation to get the percent rank, the rank of the users was divided by the number of rows,
        the result was subtracted from 1 and rounded to 2 decimal place, this will give the percentile proportion
        e.g 0.95 which is 5%, so any value between this and 0.98 is in top 5%"""
        df_grouped['activity_rank_prop'] = df_grouped['activity_rank'].apply(lambda x: round((1-(x/length_df)),2))
        df_grouped['logged_rank_prop'] = df_grouped['logged_rank'].apply(lambda x: round((1-(x/length_df)),2))
        """Here a function was used to check if the rank_prop falls within certain range"""
        def rank_percent(x):
            if x >= 0.99:
                return '1%'
            elif x >= 0.95 and x<=0.98:
                return '5%'
            elif x>= 0.90 and x<= 0.94:
                return '10%'
            elif x<= 0.93:
                return '25%'
        df_grouped['activity_percent'] = df_grouped['activity_rank_prop'].apply(rank_percent)
        df_grouped['logged_percent'] = df_grouped['logged_rank_prop'].apply(rank_percent)
        return df_grouped
        
    

In [ ]:
round(0.02,2)

0.02

In [ ]:
login = login_activity_count(df,df2)
df_grouped =login.login_activity_count()
df_grouped.head()

userid  logged_yes  activity_yes  activity_rank  logged_rank  \
0       0           0          7257            8.0        824.0   
1       1           1             2         1047.0        509.5   
2       2         169         44854            1.0          1.0   
3       3         107         12815            3.0          3.0   
4       4           0             3         1046.0        824.0   

   activity_rank_prop  logged_rank_prop activity_percent logged_percent  
0                0.99              0.21               1%            25%  
1                0.00              0.51              25%            25%  
2                1.00              1.00               1%             1%  
3                1.00              1.00               1%             1%  
4                0.00              0.21              25%            25%

In [ ]:
np.max(df_grouped['logged_yes']),np.min(df_grouped)

169

In [ ]:
log = login_activity_count(df)
log.login_activity_count()

id                     eventname component    action     target  \
0   1     \core\event\course_viewed      core    viewed     course   
1   2     \core\event\user_loggedin      core  loggedin       user   
2   3  \core\event\dashboard_viewed      core    viewed  dashboard   
3   4     \core\event\course_viewed      core    viewed     course   
4   5     \core\event\course_viewed      core    viewed     course   

  objecttable  objectid crud  edulevel  contextid  ...  contextinstanceid  \
0        None       NaN    r         2          2  ...                  1   
1        user       2.0    r         0          1  ...                  0   
2        None       NaN    r         0          5  ...                  2   
3        None       NaN    r         2          2  ...                  1   
4        None       NaN    r         2          2  ...                  1   

   userid  courseid  relateduserid  anonymous  \
0       0         1            NaN          0   
1       2         0            NaN          0   
2       2         0            2.0          0   
3       0         1            NaN          0   
4       0         1            NaN          0   

                               other timecreated  origin              ip  \
0                                 N;  1548311114     web  105.233.48.139   
1  a:1:{s:8:"username";s:5:"admin";}  1548311148     web  105.233.48.139   
2                                 N;  1548311179     web  105.233.48.139   
3                                 N;  1548313994     web  105.233.48.139   
4                                 N;  1548319150     web   41.13.254.147   

  realuserid  
0        NaN  
1        NaN  
2        NaN  
3        NaN  
4        NaN  

[5 rows x 21 columns]

In [ ]:
%%sql
SELECT *
from mdl_logstore_standard_log
limit 1;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(3995, '\\core\\event\\message_sent', 'core', 'sent', 'message', 'messages', 36, 'c', 0, 1, 10, 0, 3, 0, 56, 0, 'a:1:{s:8:"courseid";i:1;}', 1549650081, 'web', '2003:a:310:a700:566:e712:b296:c9c2', None)]

In [ ]:
%%sql 
SELECT Extract(MONTH FROM to_timestamp(firstaccess)),sum(EXTRACT(HOUR FROM to_timestamp(lastaccess-firstaccess))) as total_hours_spent
from mdl_user
where confirmed = 1 AND deleted = 0
GROUP BY EXTRACT(MONTH FROM to_timestamp(firstaccess));


 * postgresql://Samsung:***@localhost/moodle
7 rows affected.


[(3.0, 513.0),
 (7.0, 171.0),
 (1.0, 27.0),
 (2.0, 2290.0),
 (6.0, 204.0),
 (4.0, 925.0),
 (5.0, 638.0)]